File: libs/power-log-parser/src/lib/parsers/initial-card-in-deck-parser.ts
import { GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { FullEntity, Node } from '../models';
import { GameAction } from '../models/game-action';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class InitialCardInDeckParser implements ActionParser {
	readonly ParserName = 'InitialCardInDeckParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		// C# checks typeof(GameAction) — an abstract class — so this never matches
		// (the parent of initial deck FullEntities is typeof(Game), not typeof(GameAction))
		const appliesOnFullEntity =
			node.Type === FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.DECK as number) &&
			node.Parent?.Type === GameAction;
		return stateType === StateType.PowerTaskList && appliesOnFullEntity;
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const controllerId = fullEntity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'INITIAL_CARD_IN_DECK',
				GameEventHelper.CreateProvider(
					'INITIAL_CARD_IN_DECK',
					null as any,
					controllerId,
					fullEntity.Id,
					this.StateFacade,
					null,
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/InitialCardInDeckParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class InitialCardInDeckParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public InitialCardInDeckParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            var appliesOnFullEntity = node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.DECK
                // We want only the cards created when the game starts
                && node.Parent.Type == typeof(GameAction);
            return stateType == StateType.PowerTaskList
                && appliesOnFullEntity;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var controllerId = fullEntity.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                fullEntity.TimeStamp,
                "INITIAL_CARD_IN_DECK",
                GameEvent.CreateProvider(
                    "INITIAL_CARD_IN_DECK",
                    null,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    null),
                true,
                node) };
        }
    }
}
